test/hotspot/jtreg/serviceability/jvmti/vthread/VThreadEventTest/VThreadEventTest.java

```diff
@@ -0,0 +1,180 @@
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @bug 8312174
+ * @summary missing JVMTI events from vthreads parked during JVMTI attach
+ * @requires vm.continuations
+ * @requires vm.jvmti
+ * @run main/othervm/native
+ *     -Djdk.attach.allowAttachSelf=true -XX:+EnableDynamicAgentLoading VThreadEventTest attach
+ */
+
+import com.sun.tools.attach.VirtualMachine;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.locks.LockSupport;
+import java.util.List;
+import java.util.ArrayList;
+
+class Counter {
+    private int count = 0;
+
+    Counter (int count) {
+        this.count = count;
+    }
+
+    public synchronized void decr() {
+        count--;
+        notify();
+    }
+
+    public synchronized void await() {
+        try {
+            while (count > 0) {
+                wait(1);
+            }
+        } catch (InterruptedException ex) {
+            throw new RuntimeException("wait was interrupted: " + ex);
+        }
+    }
+}
+
+public class VThreadEventTest {
+    static final int TCNT1 = 10;
+    static final int TCNT2 = 4;
+    static final int TCNT3 = 4;
+    static final int THREAD_CNT = TCNT1 + TCNT2 + TCNT3;
+    static final long TIMEOUT_BASE = 1_000_000L;
+
+    private static void log(String msg) { System.out.println(msg); }
+
+    private static native int threadEndCount();
+    private static native int threadMountCount();
+    private static native int threadUnmountCount();
+
+    private static volatile int completedNo;
+    private static volatile boolean attached;
+    private static boolean failed;
+    private static List<Thread> threads = new ArrayList(TCNT1);
+
+    public static void main(String[] args) throws Exception {
+        if (Runtime.getRuntime().availableProcessors() < 8) {
+            log("WARNING: test expects at least 8 processors.");
+        }
+        Counter ready1 = new Counter(THREAD_CNT);
+        Counter ready2 = new Counter(THREAD_CNT);
+        Counter mready = new Counter(1);
+
+        try (ExecutorService executorService = Executors.newVirtualThreadPerTaskExecutor()) {
+            for (int tCnt = 0; tCnt < TCNT1; tCnt++) {
```

Comment:
Could you please add a comment before each test group creation block about expected state

Reply:
Good suggestion. Fixed now.